mountinfo.go

```diff
@@ -0,0 +1,184 @@
+func getMountInfoPartString(parts []string, idx int) string {
+	if idx >= len(parts) {
+		return ""
+	}
+	return parts[idx]
+}
+
+// Checks if the 'keys' in the optional fields in the mountinfo line are acceptable
+// Allowed 'keys' are shared, master, propagate_from, unbindable
+func isValidOptionalField(field string) bool {
+	acceptableFields := []string{"shared", "master", "propagate_from", "unbindable"}
+	for _, acceptable := range acceptableFields {
+		if field == acceptable {
+			return true
+		}
+	}
+	return false
+}
+
+// Reads each line of the mountinfo file, and returns a list of formatted MountInfo structs
+func parseMountInfo(r io.Reader) ([]*MountInfo, error) {
+	mounts := []*MountInfo{}
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		mountString := scanner.Text()
+		parsedMounts, err := parseMountInfoString(mountString)
+		if err != nil {
+			return nil, err
+		}
+		mounts = append(mounts, parsedMounts)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return mounts, err
+	}
+
+	return mounts, nil
+}
+
+// Parses a mountinfo file line, and converts it to a MountInfo struct
+// An important check here is to see if the hyphen separator, as if it does not exist,
+// it means that the line is malformed
+func parseMountInfoString(mountString string) (*MountInfo, error) {
+	var err error
+
+	// OptionalFields can be zero, hence these checks to ensure we do not populate the wrong values in the wrong spots
+	separatorIndex := strings.Index(mountString, "-")
+	if separatorIndex == -1 {
+		return nil, fmt.Errorf("no separator found in mountinfo string: %s", mountString)
+	}
+	beforeFields := strings.Fields(mountString[:separatorIndex])
+	afterFields := strings.Fields(mountString[separatorIndex+1:])
+	splits := strings.Fields(mountString)
+	if len(splits) < 7 {
+		return nil, fmt.Errorf("too few fields")
+	}
+	mount := &MountInfo{
+		MajorMinorVer:  getMountInfoPartString(beforeFields, 2),
+		Root:           getMountInfoPartString(beforeFields, 3),
+		MountPoint:     getMountInfoPartString(beforeFields, 4),
+		Options:        mountOptionsParser(getMountInfoPartString(beforeFields, 5)),
+		OptionalFields: nil,
+		FSType:         getMountInfoPartString(afterFields, 0),
+		Source:         getMountInfoPartString(afterFields, 1),
+		SuperOptions:   mountOptionsParser(getMountInfoPartString(afterFields, 2)),
+	}
+
+	mount.MountId, err = strconv.Atoi(getMountInfoPartString(beforeFields, 0))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse mount ID")
+	}
+	mount.ParentId, err = strconv.Atoi(getMountInfoPartString(beforeFields, 1))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse parent ID")
+	}
+	// Has optional fields, which is a space separated list of values
+	// Example: shared:2 master:7
+	if len(beforeFields) > 6 {
+		mount.OptionalFields = make(map[string]string)
+		optionalFields := beforeFields[6:]
+		for _, field := range optionalFields {
+			optionSplit := strings.Split(field, ":")
```

Comment:
Since Split/SplitN should always return at least length 1, I think you can simplify this section a bit to something like this.
```golang
optionSplit := strings.SplitN(field, ":", 2)
value = ""
if len(optionSplit) == 2 {
	value = optionSplit[1]
}
if isValidOptionalField(optionSplit[0]) {
	mount.OptionalFields[optionSplit[0]] = value
}
```